### PR TITLE
rpc: Add submit option to generateblock

### PR DESCRIPTION
**Pull request by maflcko**
When submit is turned off, a block can be generated and returned as hex, to be used for further tests. For example, it can be submitted on a different node, on a different interface (like p2p), or just never submitted and be used for other testing purposes.

### CODE REVIEW

**Comment by MarcoFalke** [comment]
Requested by @instagibbs, I believe, in comment https://github.com/bitcoin/bitcoin/pull/17693#discussion_r393727104

This work is based on the `generateblock` RPC added by @andrewtoth in #17693

**Comment by andrewtoth** [comment]
`generatetodescriptor` will be available for at least 2 releases. Is there any concern that other users will depend on it? I suppose workarounds for it are easy enough.

File: doc/release-notes.md [hunk 1]
```diff
@@ -104,6 +104,9 @@ Low-level changes
 Tests
 -----
 
+- The `genereratetodescriptor` RPC has been removed and replaced by the
```
Comment (jonatack) [inline doc/release-notes.md:107]: s/genereratetodescriptor/generatetodescriptor/

**Comment by MarcoFalke** [comment]
If anyone was quick enough to add `generatetodescriptor` to their test scripts, they are surely quick enough to pull it out again the next time they update their node to a new major version.

**Review by jonatack — COMMENTED** [review]
Does removing the generatetodescriptor RPC require a deprecation cycle?

**Comment by MarcoFalke** [comment]
> Does removing the generatetodescriptor RPC require a deprecation cycle?

I'd say no, because it is only used for testing. See also https://github.com/bitcoin/bitcoin/pull/18933#issuecomment-626356218

File: doc/release-notes.md (hunk 1 above)
Comment (MarcoFalke) [reply doc/release-notes.md]: Thanks, fixed.

**Comment by andrewtoth** [comment]
I haven't used `generatetodescriptor` myself, but I would imagine a use case would be to mine transactions in the mempool with it. With `generateblock` a user has to manually add all `txid`s to get the same functionality. If we are replacing it, should we also add an option to `generateblock` to mine the mempool?

**Comment by MarcoFalke** [comment]
Ok, that can be done as a follow-up. Dropped the last commit faefc219b63825b3393ccea683645ecbbac0e169 because it was too controversial.

File: src/rpc/mining.cpp [hunk 2]
```diff
@@ -1187,7 +1200,7 @@ static const CRPCCommand commands[] =
 
     { "generating",         "generatetoaddress",      &generatetoaddress,      {"nblocks","address","maxtries"} },
     { "generating",         "generatetodescriptor",   &generatetodescriptor,   {"num_blocks","descriptor","maxtries"} },
-    { "generating",         "generateblock",          &generateblock,          {"output","transactions"} },
+    { "generating",         "generateblock",          &generateblock,          {"output","transactions","submit"} },
```
Comment (instagibbs) [inline src/rpc/mining.cpp:1203]: Also should be added to "src/rpc/client.cpp" to be parsed as non-string on -cli, right?

**Review by instagibbs — COMMENTED** [review]
ACK aside from the json question

File: src/rpc/mining.cpp (hunk 2 above)
Comment (MarcoFalke) [reply src/rpc/mining.cpp:1203]: Doh. Can't wait until RPCMan does that for me

#18531

Comment (instagibbs) [reply src/rpc/mining.cpp:1203]: oh nice, will review

**Comment by andrewtoth** [comment]
The comment this is based on (https://github.com/bitcoin/bitcoin/pull/17693#discussion_r393727104) suggests to bypass validity rules. However, https://github.com/bitcoin/bitcoin/pull/18933/files#diff-9651347c8e00bed3ddc7631de569406dL364 still does that. Do we want to skip that check as well?

File: src/rpc/mining.cpp (hunk 2 above)
Comment (MarcoFalke) [reply src/rpc/mining.cpp]: Thanks, fixed.

**Comment by MarcoFalke** [comment]
Rebased and added missing hunk for client.cpp

**Comment by MarcoFalke** [comment]
Rebased

**Comment by MarcoFalke** [comment]
rebased

**Comment by achow101** [comment]
Closing this as it has not had any activity in a while. If you are interested in continuing work on this, please leave a comment so that it can be reopened.

**Review by sedited — APPROVED** [review]
tACK fa18504d5767a40dc9827fb081633219bf251001

**Pull request merged**